From: Markus Beth <markus.beth@web.de>
Date: Wed, 22 May 2019 19:51:40 +0200
Subject: fixed version number and typo

---
 doc/ChangeLog | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/doc/ChangeLog b/doc/ChangeLog
index 3a054d1..62eb5ef 100644
--- a/doc/ChangeLog
+++ b/doc/ChangeLog
@@ -1,4 +1,4 @@
-rdesktop (1.8.5)
+rdesktop (1.8.6)
   * Fix protocol code handling new licenses
 
   -- Pierre Ossman <ossman@cendio.se> 2019-05-16
@@ -81,7 +81,7 @@ rdesktop (1.8.1)
   * Fix issue with reconnect, make use of deactivate variable
   * Added 4 new disconnect reasons with exit codes
   * Fix issues of window handling in SeamlessRDP parts of rdesktop
-  * Fix a backward compability with OpenSSL < 0.9.9
+  * Fix a backward compatibility with OpenSSL < 0.9.9
   * Fix a bug when code needs a x window available but there are none.
   * Fix a sigsegv zeroing memory
   * Fix a 64bit portability issue
