Description: add missing format string to snprintf's
Last-Update: <2012-02-08>

--- rdesktop-1.7.1.orig/rdesktop.c
+++ rdesktop-1.7.1/rdesktop.c
@@ -1584,7 +1584,7 @@ load_licence(unsigned char **data)
 	if (home == NULL)
 		return -1;
 
-	snprintf((char*)hi, 16, g_hostname);
+	snprintf((char*)hi, 16, "%s", g_hostname);
 	sec_hash_sha1_16(ho, hi, g_static_rdesktop_salt_16);
 	sec_hash_to_string(hash, 40, ho, 22);
 
@@ -1634,7 +1634,7 @@ save_licence(unsigned char *data, int le
 		return;
 	}
 
-	snprintf((char*)hi,16,g_hostname);
+	snprintf((char*)hi,16,"%s",g_hostname);
 	sec_hash_sha1_16(ho, hi, g_static_rdesktop_salt_16);
 	sec_hash_to_string(hash, 40, ho, 20);
 
